docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,365 @@
+  
+  // Type is a field that specifies the type of out-of-band operator task to be performed. 
+  // It can have values of all supported operations/tasks type eg. "OnDemandSnapshotTask" etc.
+  Type string `json:"taskType"`
+
+  // Config is a task specific configuration.
+  Config *runtime.RawExtension `json:"taskConfig,omitempty"`
+
+  // TTLSecondsAfterFinished is the time-to-live after which the task and 
+  // related resources will be garbage collected.
+  // +optional
+  TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the `Status` (current state) of the EtcdOperatorTask custom resource as the custom resource's `Status` will be used to monitor the progress of the task.
+
+* To capture the Status of a task, a field `taskStatus` is defined in status.
+* If operation involves many stages, so to capture the status of intermediate or any stage, `.status.lastOperation` will be useful.
+* To capture the all last occured errors on resource for a task, a field `lastErrors` is defined in status.
+
+```go
+// TaskState represents the state of the task.
+type TaskState string
+
+const (
+  Failed TaskState = "failed"
+  Pending TaskState = "pending"
+  Rejected TaskState = "rejected"
+  Completed TaskState = "completed"
+  InProgress TaskState = "inProgress"
+)
+
+// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
+type EtcdOperatorTaskStatus struct {
+  // ObservedGeneration is the most recent generation observed for the resource.
+  ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+  // Status of the last operation
+  Status TaskState `json:"taskStatus"`
+  // Time at which operation is been triggered.
+  InitiatedAt metav1.Time `json:"initiatedAt"`
+  // LastError represents the errors when processing the task.
+  LastErrors []LastError `json:"lastErrors,omitempty"`
+  // Captures the last operation
+  // +optional
+  LastOperation *LastOperation `json:"lastOperation,omitempty"`
+  // Conditions represents the latest available observations of an object's current state.
+  // +optional
+  Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type LastOperation struct {
+  // Name of the LastOperation.
+  Name opsName `json:"name"`
+  // Status of the last operation, one of pending, progress, completed, failed.
+  State OperationState `json:"state"`
+  // Last time the operation state transitioned from one to another.
+  LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+  // A human readable message indicating details about the last operation.
+  Reason string `json:"reason"`
+}
+
+```
+
+### Lifecycle
+
+#### Creation
+
+Task(s) can be created by creating an instance of the `EtcdOperatorTask` custom resource specific to a task.
+
+> Note: For future improvements, a creation of an instance of the `EtcdOperatorTask` custom resource can be done via CLI wrapper.
```

Comment:
```suggestion
> Note: In future, either a `kubectl` extension plugin or a `druidctl` will be introduced. Dedicated sub-commands will be created for each `out-of-band` task. This will drastically increase the usability for an operator for such tasks as the CLI extension will automatically created relevant instance of `EtcdOperatorTask` with the provided configuration.
```